### PR TITLE
proxy: cache range request in proxy

### PR DESCRIPTION
**Pull request by mqliang**
Did not try to full implement this, This pr is just for reviewing and discussion: 
1) use groupcache to support cache invalidation
2) this cache could cache response for a single key.

@xiang90  To cache Range request, it seems that we must write an interval tree with max leaf numbers and lru? Or maybe you have any better ideas?

### CODE REVIEW

File: proxy/grpcproxy/cache/store.go [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+func defaultCompactor(deltas Deltas, revision uint64) Deltas {
+	index := sort.Search(len(deltas), func(i int) { return deltas[i] > revision })
+	ret := Deltas{}
+	ret = append(ret, deltas[index:])
+	return ret
+}
+
+// Deltas is a list of one or more 'Delta's to an individual object.
+// The oldest delta is at index 0, the newest delta is the last one.
+type Deltas []Delta
```
Comment (xiang90) [inline proxy/grpcproxy/cache/store.go]: I think this is a little bit over kill. We already have lru there. We can just let lru take care of everything at day 1, and make it more sophisticated later when we know more about the use case and expectation from users.

So my suggestion is:

1. use key + revision as the cache key in lru.
2. add another map that maps from a key to its most recent revision in lru.

When users get key with a specific revision, we can query the lru cache directly.

When users get key without a specific revision, we find the revision from the map and then query lru.

**Comment by mqliang** [comment]
@xiang90  An interval tree with max leaf numbers and lru seems too complicated to me. We could also just cache response for Range request, instead of Get request(Get request is just a particular Range request). but it may waste some memory, since two Range may overlapping.

**Comment by xiang90** [comment]
@mqliang We already have an interval rb-tree implementation in etcd: see https://github.com/coreos/etcd/blob/master/pkg/adt/interval_tree.go.

However, I think we can keep this simple today by just index the range query by key_endkey_revision, just as normal get query. Yes, if someone queries a sub-range, we will end up with double caching. But it is simple. We can start with this simple thing and see how it goes? /cc @heyitsanthony  idea?

**Comment by xiang90** [comment]
@mqliang For interval tree thing, see my previous last comment. So yes, I agree. We can always make this better later. The cache is an implementation detail right now. The more important thing is probably to define the cache. What is the guarantee of the cache? For query with revision, we can easily say it always return the fresh data since the version is there. For query without revision, what is the freshness guarantee? 

I would suggest it to be time-based invalidation. We can make it watch-based later.

**Comment by mqliang** [comment]
@xiang90 
Starting from simple stuff and then iterate sounds good to me

> I would suggest it to be time-based invalidation

How about lru AND time-based. It's not difficult IMO: we could define a expire time when we insert a response into the cache. And the expiration could be lazy: if cache hit, but it already expired, then invalidate the cache and report cache missing. One benefit of lru is that user could configure the size of cache so that the cache will not use too many memory.

**Comment by xiang90** [comment]
@mqliang 

Ah. I do not think we need to evict anything from lru ourselves. If you are happy about my in-line comment, we only need to clean up the map from key to its most recent revision. And yes, it can be a lazy removal.

**Comment by heyitsanthony** [comment]
@xiang90 I'm not too concerned about double caching costs. On the other hand, reasoning about revisions and queries on a per key basis looks tricky (e.g., interpolating on Get(WithRev()) seems broken with the current code).

It's probably better to have something simple that obviously works (e.g., caching only exact request -> response, maybe with lru, maybe periodic cache drop) running first before getting smart about it.

**Comment by xiang90** [comment]
> It's probably better to have something simple that obviously works (e.g., caching only exact request -> response, maybe with lru, maybe periodic cache drop) running first before getting smart about it.

So you suggest that we do read-through if the revision is not provided?

**Comment by heyitsanthony** [comment]
@xiang90 rev=0 is still a request so it will be cached. The store revisions will be all over the place for rev=0, but the responses would be valid.

Next iteration could partition rev=0 and rev!=0 caches and have some way to update/expire rev=0 requests.

**Comment by mqliang** [comment]
@xiang90 @heyitsanthony  cache the whole request->response sounds good to me. And I have several questions:

1) Do we plan to support run etcdproxy in standalone mode in long run? If we do, then etcdproxy should be compatible with the client right? 
2) Dose the compaction operation ensure atomic？
* If compaction is atomic(store stays unchanged if compaction failed, instead of part of the keys are compacted, and others not), when etcdproxy forward the request to etcd server, it should compact the local cache if server response succeeded and should not compact the local cache if server response failed. 
* If compaction is not atomic, we should always compact the local cache, even if server side response failed.

**Comment by xiang90** [comment]
> 1) Do we plan to support run etcdproxy in standalone mode in long run? If we do, then etcdproxy should be compatible with the client, right? 

Yes. The compatibility happens at API layer though. We use the same gRPC API as etcdserver.

>  Dose the compaction operation ensure atomic？

Yes. But from client point of view, even if we do not compact proxy's cache but compact the server, it does not really matter I think.

File: proxy/grpcproxy/cache/store.go [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+	if revision > c.latestCompactionRev {
+		c.latestCompactionRev = revision
+	}
+}
+
+// deepHashObject writes specified object to hash using the spew library
+// which follows pointers and prints actual values of the nested objects
+// ensuring the hash does not change when a pointer changes.
+// This function is borrowed from kubernetes project.
+func deepHashObject(hasher hash.Hash, objectToWrite interface{}) {
```
Comment (xiang90) [inline proxy/grpcproxy/cache/store.go]: can we get rid of this thing? there are not a lot of fields in https://github.com/coreos/etcd/blob/master/etcdserver/etcdserverpb/rpc.proto#L151-L193. Let's just hash it ourselves.

File: proxy/grpcproxy/cache/store.go [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	if req.Revision > c.latestCompactionRev {
+		c.lru.Add(key, resp)
+	}
+}
+
+// Get looks up the caching response for a given request.
+func (c *cache) Get(req *pb.RangeRequest) (*pb.RangeResponse, bool) {
```
Comment (xiang90) [inline proxy/grpcproxy/cache/store.go]: probably we want to return error for compacted. so that request will not propagate to etcd.

**Comment by mqliang** [comment]
@xiang90 I try to import the groupcache package and update Godeps/vendor, but seems something goes wrong. Mind to help take a look?

```
➜  etcd git:(cache) ✗ go env
GOARCH="amd64"
GOBIN=""
GOEXE=""
GOHOSTARCH="amd64"
GOHOSTOS="darwin"
GOOS="darwin"
GOPATH="/Users/mqliang/Documents/NetEase/Etcd/Godeps:/Users/mqliang/Documents/NetEase/Etcd"
GORACE=""
GOROOT="/usr/local/go"
GOTOOLDIR="/usr/local/go/pkg/tool/darwin_amd64"
GO15VENDOREXPERIMENT="1"
CC="clang"
GOGCCFLAGS="-fPIC -m64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fno-common"
CXX="clang++"
CGO_ENABLED="1"
```

```
➜  etcd git:(cache) ✗ pwd
/Users/mqliang/Documents/NetEase/Etcd/src/github.com/coreos/etcd
➜  etcd git:(cache) ✗ cd cmd 
➜  cmd git:(cache) ✗ godep restore
➜  cmd git:(cache) ✗ cd ../
➜  etcd git:(cache) ✗ go get github.com/golang/groupcache/lru
➜  etcd git:(cache) ✗ rm -rf Godeps vendor
➜  etcd git:(cache) ✗ ln -s cmd/vendor vendor
➜  etcd git:(cache) ✗ godep save ./...

godep: Package (golang.org/x/tools/blog/atom) not found

```

**Comment by xiang90** [comment]
@mqliang Try ./script/updatedep.sh instead?

**Comment by mqliang** [comment]
@xiang90 Same result:
```
➜  etcd git:(cache) ✗ sh -x ./scripts/updatedep.sh 
+ rm -rf Godeps vendor
+ ln -s cmd/vendor vendor
+ godep save ./...
godep: Package (golang.org/x/tools/blog/atom) not found
+ rm -rf cmd/Godeps
+ rm vendor
+ mv Godeps cmd/
mv: rename Godeps to cmd/Godeps: No such file or directory
```

It's really strange, I meet the same problem last time when I try to import tabwriter package.

**Comment by xiang90** [comment]
after you do godep restore, is `golang.org/x/tools/blog/atom` in your gopath?

**Comment by mqliang** [comment]
@xiang90 
> after you do godep restore, is golang.org/x/tools/blog/atom in your gopath?

Unfortunately, no. There is no the whole golang.org/x/tools actually

**Comment by gyuho** [comment]
@mqliang Yeah currently Godep is quite broken with etcd. Workaround is just move all your GOPATH for the moment to somewhere else, and `godep restore` to get the clean dependencies and then get the new dependencies, and godep save.

I am planning to play with glide to see if we can make this process easier.

**Comment by xiang90** [comment]
@gyuho What you suggested is what he did i think. Can you verify that it actually works for you now?

**Comment by gyuho** [comment]
Ok, I will try to `godep` from his branch.

File: proxy/grpcproxy/cache/store.go [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+const DefaultMaxEntries = 2048
+
+type Cache interface {
+	Add(req *pb.RangeRequest, resp *pb.RangeResponse)
+	Get(req *pb.RangeRequest) (*pb.RangeResponse, bool, error)
+	Compact(revision int64)
+}
+
+// keyFunc returns the key of an request, which is used to look up in the cache for it's caching response.
+func keyFunc(req *pb.RangeRequest) uint64 {
```
Comment (xiang90) [inline proxy/grpcproxy/cache/store.go]: let's do this instead

``` go
keyFunc(req *pb.RangeRequest) string {
    // todo: use marshalTo to reduce allocation
    b, err := req.Marshal()
    if err != nil {
        panic(err)
    }
    return string(b)
}
```

Although this costs more space to save the key, but this is loss-less and free of conflicts. There might be conflict using hash32, and we have to check the conflict.

File: proxy/grpcproxy/cache/store.go [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+	if req.Revision > c.latestCompactionRev {
+		c.lru.Add(key, resp)
+	}
+}
+
+// Get looks up the caching response for a given request.
+func (c *cache) Get(req *pb.RangeRequest) (*pb.RangeResponse, bool, error) {
+	key := keyFunc(req)
+
+	// NODE: we use WLock instead of Rlock here since we change lru's state in Get() method.
```
Comment (xiang90) [inline proxy/grpcproxy/cache/store.go]: remove this note? we have a lru.remove in the code path. we need lock anyway.

File: proxy/grpcproxy/cache/store.go [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+	return &cache{
+		lru: lru.New(maxCacheEntries),
+	}
+}
+
+// cache implements Cache
+type cache struct {
+	lock                sync.RWMutex
+	lru                 *lru.Cache
+	latestCompactionRev int64
```
Comment (xiang90) [inline proxy/grpcproxy/cache/store.go]: compactedRev

File: proxy/grpcproxy/cache/store.go [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+}
+
+// Add adds response for a request to the cache.
```
Comment (xiang90) [inline proxy/grpcproxy/cache/store.go]: Add adds the response of a request to the cache if its revision is larger than the compacted revision of the cache.

File: proxy/grpcproxy/cache/store.go [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	if req.Revision > c.latestCompactionRev {
+		c.lru.Add(key, resp)
+	}
+}
+
+// Get looks up the caching response for a given request.
```
Comment (xiang90) [inline proxy/grpcproxy/cache/store.go]: Get is also responsible for lazy eviction when accessing compacted entries.

File: proxy/grpcproxy/cache/store.go [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+	}
+
+	if resp, ok := c.lru.Get(key); ok {
+		return resp.(*pb.RangeResponse), ok, nil
+	}
+	return nil, false, nil
+}
+
+// Compact invalidate all caching response before the given rev.
+// Note: the invalidation is lazy
```
Comment (xiang90) [inline proxy/grpcproxy/cache/store.go]: Remove `Note:` replace with The invalidation is lazy. The actual removal happens when the entries is accessed.

File: proxy/grpcproxy/kv.go [hunk 10]
```diff
@@ -17,22 +17,30 @@ package grpcproxy
 import (
 	"github.com/coreos/etcd/clientv3"
 	pb "github.com/coreos/etcd/etcdserver/etcdserverpb"
-
+	"github.com/coreos/etcd/proxy/grpcproxy/cache"
```
Comment (xiang90) [inline proxy/grpcproxy/kv.go]: this should group with the first two imports. (all internal imports)

File: proxy/grpcproxy/cache/store.go [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	if req.Revision > c.latestCompactionRev {
+		c.lru.Add(key, resp)
+	}
+}
+
+// Get looks up the caching response for a given request.
+func (c *cache) Get(req *pb.RangeRequest) (*pb.RangeResponse, bool, error) {
```
Comment (xiang90) [inline proxy/grpcproxy/cache/store.go]: let's return just resp and error. we can check err compacted in the outside to decide if we should forward to etcdserver.

File: proxy/grpcproxy/kv.go [hunk 12]
```diff
@@ -17,22 +17,30 @@ package grpcproxy
...
 func NewKvProxy(c *clientv3.Client) *kvProxy {
 	return &kvProxy{
-		c: c,
+		c:     c,
+		cache: cache.NewCache(cache.DefaultMaxEntries),
 	}
 }
 
 func (p *kvProxy) Range(ctx context.Context, r *pb.RangeRequest) (*pb.RangeResponse, error) {
+	if resp, exist, err := p.cache.Get(r); err != nil || (err == nil && exist) {
```
Comment (xiang90) [inline proxy/grpcproxy/kv.go]: resp, err := p.cache.Get(r)

if resp != nil || err == rpctypes.ErrGRPCCompacted {
    return resp, err
}

File: proxy/grpcproxy/cache/store.go [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+
+	// NODE: we use WLock instead of Rlock here since we change lru's state in Get() method.
+	// (remove the key from cache if it's revision has been compacted; update least recently usage information)
+	// So, we can not call Get() method concurrently.
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	if req.Revision > c.latestCompactionRev {
+		c.lru.Remove(key)
+		return nil, false, mvcc.ErrCompacted
```
Comment (xiang90) [inline proxy/grpcproxy/cache/store.go]: use `rpctypes.ErrGRPCCompacted `

**Comment by xiang90** [comment]
@mqliang I am not super worry about unit test for know. However, can you try to setup an etcd cluster with this cache enabled to make sure it works as expected. Thanks!

File: proxy/grpcproxy/cache/store.go (hunk 13 above)
Comment (xiang90) [inline proxy/grpcproxy/cache/store.go]: ideally define cache.ErrCompacted and convert this at grpcProxy layer...

**Comment by xiang90** [comment]
@mqliang Are you still working on this one?

**Comment by mqliang** [comment]
@xiang90 Will update this soon. 

I was thinking what should I do in the next step. Maybe: proxy watch the whole store from cluster member(by watching the root dir) so that make the local cache up to date. So the proxy could server all read requests and watch requests from local cache directly, without considering cache hit/miss?

BTW, should we add cache for the v2 proxy(the http_proxy)?

**Comment by mqliang** [comment]
@xiang90 How can I bring up a local v3 etcd cluster? I run `goreman -f Procfile start ` but it seems that what I got was a v2 cluster(since etcdctlV3 can not work, but etcdctlV2 works fine). 

BTW, how etcd server side decide run in v2 mode or v3 mode? I looks through the code, etcdctl decide v2/v3 by a "ETCDCTL_API" environment variable, but it seems etcd server will always run in v2 mode, (see https://github.com/coreos/etcd/blob/master/etcdmain/main.go#L36).

**Comment by xiang90** [comment]
try current master. it will bring up a etcd server with v2 and v3. The code you refer is correct. It brings up etcd server or proxyv2.

**Comment by xiang90** [comment]
> I was thinking what should we do in the next step. Maybe: proxy watch the whole store from cluster member(by watching the root dir) to make the local cache up to date. So the proxy could server all read requests and watch requests from local cache directly, without considering cache hit/miss?

For this, you can simply bring up an etcd cluster and use mirror-make to mirror the key-space you want. 

We can use https://github.com/coreos/etcd/blob/master/clientv3/mirror/syncer.go to implement this easily in proxy too. But this is probably a future step.

>  BTW, should we add cache for the v2 proxy(the http_proxy)?

We do not have any plan at our side. v2 proxy does not try to understand etcd request. It is a transparent gateway. probably you will need more work if you want to do it.

**Comment by mqliang** [comment]
@xiang90 I may have falsely thinking the significance of etcd-proxy is read/write splitting: serve read requests from proxy local cache, forward write requests to cluster member. 

If we already have etcd-mirror, why do we need etcd-proxy? Maybe etcd-proxy should works as a middleware for request forwarding: forward serializable read request and watch request to mirror, and forward other requests to cluster member?

**Comment by xiang90** [comment]
@mqliang The mirror-etcd does not forward write. It is just a etcd server linked by a mirro-maker to another etcd server. It is a different thing than the proxy we are talking about here.

The responsibility of proxy is very clear:

1. cache read
2. write through write
3. colease watch
4. potentially namespace based sharding
5. watch based cache of a hot namespace

What you try to add into the cache is 5. And it is a longer term goal. We do not want to do it from day1.

**Comment by mqliang** [comment]
@xiang 

It seems we can make proxy a transparent layer which is dedicated to request forwarding(and potentially sharding in the future). If proxy/cluster member/mirror cooperate as following to implement read/write splitting, why should we need a cache in proxy layer?

```
					  +---------+
				      |	 client	|
					  +---------+
							|			
							V
			+------------------------------+
			|			etcd-proxy         |	
			+------------------------------+
				|						|
		write / |						| serializable read /
	quorum-read |						| watch requests
	requests	|						|	
				|						|
				V						V
+-------------------+		sync	 +--------+
|	cluster-member	|--------------->| mirror |		
+-------------------+				 +--------+
```

**Comment by xiang90** [comment]
@mqliang Mirror is stateful (it is an etcdserver anyway) and has its cost. Proxy is stateless and easy to swap in and out. I would image to see most people just set up 10s of proxies and will handle their read/watch good enough. For the model you described, it is useful when you have a enough larger deployment or when sharding comes into play. These are all future plan. We now need to get the stateless proxy with caching/watch coalescing done. Then we can iterate on it.

**Comment by mqliang** [comment]
> xiang90
> Mirror is stateful (it is an etcdserver anyway) and has its cost

Do you mean mirror will also write WAL log, make snapshot, etc. But proxy just sync store in it's RAM?

**Comment by xiang90** [comment]
> Do you mean mirror will also write WAL log, make snapshot, etc. But proxy just sync store in it's RAM?

Yes. Mirror is an full feature etcdserver. Proxy is just a stateless thing.

**Comment by xiang90** [comment]
@mqliang Shall we resolve the merge conflicts and get this merged first? Then we can start to iterate on it?

**Comment by mqliang** [comment]
@xiang90 Seems that we can not bring up a v3 proxy now, the [current code](https://github.com/coreos/etcd/blob/master/etcdmain/etcd.go#L402) only start a v2 http_proxy.

**Comment by xiang90** [comment]
@mqliang Yes. You need to write a simple driver to bring it up. I can do it later this week after we get this pr merged.

**Comment by mqliang** [comment]
cc/ @jerryzh168

**Pull request merged**